### ISSUE TITLE
'Network connection isn't performing well' message with new router

### ISSUE THREAD

**Issue by sixcuda** [moonlight-stream/moonlight-qt#714]
**Describe the bug**
When launching a game from moonlight on my steam link hardware, I receive the message, "Your network connection isn't performing well. Reduce your video bitrate setting or try a faster connection." And games are unplayable

This is happening after upgrading to a new Asus RT-AX58U router, which both my Steam Link hardware and computer are hardwired into. 

I have no problems connecting and playing games natively through the steam link, so it seems to be a problem that's isolated to moonlight and/or my new router. I had no issues with moonlight with my previous router, which was an Asus RT-N66U

Things I have tried that did not work:

- Updating firmware on router to latest
- Changing any relevant settings within the router (ie confirmed qos disabled)
- Using different network adapters on pc (my laptop has a Killer e2400 gigabit, and i've tried an external tp-link gigabit usb 3.0 ethernet adapter)
- Uninstalling/re-installing network drivers
- Different ethernet cables 

What has worked:

- I came across [a similar issue](https://github.com/moonlight-stream/moonlight-android/issues/79) in my search where a workaround suggests changing the "Speed & Duplex" setting of the pc network adapter from "auto negotiation" to "100 Mbps Full Duplex" which did resolve the issue, but I would rather not downgrade the network speed of my pc
- If I instead plug my steam link into a network switch first, and not into the router directly, this resolves the issue and is how I am currently operating. 

My networking knowledge is very limited, but it would seem to me that the switch is able to take the slower 100 Mbps connection from the steam link and somehow make it compatible with the router.

What I am ultimately hoping to determine is whether the problem is with Moonlight, or my new router. Or something else? 

The router seems to work fine with all other devices on the network, including using the steam link without moonlight. 

Any ideas would be appreciated

**Steps to reproduce**
Steam link connected directly to RT-AX58U router

**Screenshots**
none

**Affected games**
Forza Horizon 5
Ark: survival evolved
Opening steam big picture within moonlight seems to works fine, but games are still unplayable

**Other Moonlight clients**
have not tested

**Moonlight settings (please complete the following information)**
I believe all settings are default except:
1080p @ 60fps
Bitrate: 22
V-Sync is off

I don't remember what the default settings were, but these settings worked fine through my old RT-N66U router

**Gamepad-related issues (please complete if problem is gamepad-related)**
n/a

**Client PC details (please complete the following information)**
 - Steam Link Hardware: Build 822
 - Moonlight Version: 3.1.4

**Server PC details (please complete the following information)**
 - OS: Windows 10 Home 21H2
 - GeForce Experience version: 3.24.0.126
 - Nvidia GPU driver: 496.49
 - Antivirus and firewall software: Windows Defender and Windows Firewall

**Moonlight Logs (please attach)**
not sure how to produce logs from steam link

**Additional context**
none

**Comment by cgutman** [comment]
Yeah, this can happen because some switches and NICs can't handle the bursts of network traffic generated from GFE each frame. Your workarounds are pretty much the best known solutions: either use a switch that can handle it, or lower your PC to 100Mb.

**Comment by Castlenock** [comment]
Just logging in to quickly say I've had the same issue on three separate networks with my Samsung Galaxy tab lines (s7+, S8).  For my current network (high end Asus like the OP, two layer switch) I'm not reading any red flags on network traffic.  

It's dropping 24% even when I throttle to 75Mbps.  I have no issues on Wifi 6 at 150Mbps other than the jitter that comes with wifi connections...

I'll try and get more details if you need them cgutman (your work is amazing, Moonlight has transformed the way I work and play) - too many plates to balance at the moment to do an info dump, but it doesn't feel like a switch/router issue, namely in that PC to PC moonlighting at 150Mbps speeds have always been easy peasy lemon squeezy on the same networks I tried the Galaxy tabs on.

**Comment by floxigen** [comment]
Ive the same issue with a 2.5 GIG NIC

**Comment by cgutman** [comment]
> It's dropping 24% even when I throttle to 75Mbps. I have no issues on Wifi 6 at 150Mbps other than the jitter that comes with wifi connections...

How are you throttling? Moonlight's bandwidth slider won't affect this issue, because GeForce Experience still sends the frame out as fast as possible. Even if the average bandwidth is less than your connection's maximum, these blasts of traffic every 16ms can overwhelm buffers in switches and clients. 

> it doesn't feel like a switch/router issue, namely in that PC to PC moonlighting at 150Mbps speeds have always been easy peasy lemon squeezy on the same networks I tried the Galaxy tabs on.

That probably points more towards buffering behavior in the Galaxy Tab's WiFi adapter driver. A similar issue was noted in #787 with the r8169 driver on Linux.

> Ive the same issue with a 2.5 GIG NIC

If you set it to 1Gbps, does it go away?

**Comment by Castlenock** [comment]
"How are you throttling? Moonlight's bandwidth slider won't affect this issue, because GeForce Experience still sends the frame out as fast as possible. Even if the average bandwidth is less than your connection's maximum, these blasts of traffic every 16ms can overwhelm buffers in switches and clients."

Ah, given that explanation I wasn't throttling at all, just moving Moonlight's slider.  Does that buffer extend to USB NICs?  I have tried 5 different NICs (had a lot left over from work), even one that maxed out at 100mbps for giggles.  

"That probably points more towards buffering behavior in the Galaxy Tab's WiFi adapter driver. A similar issue was noted in https://github.com/moonlight-stream/moonlight-qt/issues/787 with the r8169 driver on Linux."

I'm not as knowledgable as you but I think it more to do with how the Galaxy deals with network data via WiFi adapter and when it shoves it through Ethernet.  On the S7+ before I traded it in for the S8U I observed other Ethernet NIC issues - easiest indicator I could give beyond moonlight is was a speedtest on a 1G down 40mb up in that the S7+ never went over 250 or so mbps regardless of what USB NIC I used (would be higher on wifi).  For the S8U, wired speedtests run the same as my wired rig, but whatever alterations they did on the network stack via USB NIC probably weren't enough to deal with the requirements of Moonlight.  

Appreciate the thoughts on this.

**Comment by cgutman** [comment]
> Ah, given that explanation I wasn't throttling at all, just moving Moonlight's slider. Does that buffer extend to USB NICs? I have tried 5 different NICs (had a lot left over from work), even one that maxed out at 100mbps for giggles.

Yeah, it does affect USB NICs perhaps more so than PCIe ones. If you set your host PC's NIC to 100 Mbps Full Duplex, do the frame drops go away?

**Comment by Castlenock** [comment]
"If you set your host PC's NIC to 100 Mbps Full Duplex, do the frame drops go away?"

Will try this and get back to you within the next few days.

**Comment by floxigen** [comment]
> > It's dropping 24% even when I throttle to 75Mbps. I have no issues on Wifi 6 at 150Mbps other than the jitter that comes with wifi connections...
> 
> How are you throttling? Moonlight's bandwidth slider won't affect this issue, because GeForce Experience still sends the frame out as fast as possible. Even if the average bandwidth is less than your connection's maximum, these blasts of traffic every 16ms can overwhelm buffers in switches and clients.
> 
> > it doesn't feel like a switch/router issue, namely in that PC to PC moonlighting at 150Mbps speeds have always been easy peasy lemon squeezy on the same networks I tried the Galaxy tabs on.
> 
> That probably points more towards buffering behavior in the Galaxy Tab's WiFi adapter driver. A similar issue was noted in #787 with the r8169 driver on Linux.
> 
> > Ive the same issue with a 2.5 GIG NIC
> 
> If you set it to 1Gbps, does it go away?

Yeah it does work when I switch to 1 Gbps

**Comment by belegdol** [comment]
I am seeing a similar issue while streaming from a PC connected via a cat6a cable to Fire TV Stick 4K. Anything above 50 Mbit/s causes hiccups, 50 Mbit/s sometimes gets choppy as well.

**Comment by Gremious** [comment]
I'm trying to wrap my head around this problem and I have a question:

My setup is PC -> Switch -> Router -> Switch -> Pi 5

I face this issue at 1Gb/s link negotiation speed and it does fix when dropped down to 100 Mb/s.

However, if I disconnect (just) the switch from the Pi, and plug it to the router directly it works at 1Gb/s.
I guess that switch can't handle it, so I should get a better one.
However, I was wondering why the one between my PC and router works okay?
Both switches are the same, 1G TP-Links.

---

> either use a switch that can handle it

Are there any known switches that work well?
It seems from the replies that if I have a 1Gb/s network connection any 2.5G switch should just work, as the issue is going over the 1G limit in a burst?

(@cgutman  You seem to understand the issue well, so I hope you don't mind the ping.)

**Comment by renaudcerrato** [comment]
Just tested with a USB NIC on client side, same behavior. I don't quite understand why the NIC drivers would drop ethernet frames on its own at 30+ Mbps only since the decoding pipeline is fast enough (can decode 300+ Mbps when host is downgraded to 1Gbe), there's no bottleneck here. 

The ethernet frames should be received and transfered (using DMA/IRQ) to RAM fast enough. Moreover, flow control is built-in at several layers to prevent such situation. 

iperf2 is working flawlessly too, only Moonshine seems to be affected.

Any thoughts or suggestions to dig further?

**Comment by bmlrX3k** [comment]
Hi guys, thank you very much, since I changed my network card from 2.5Gbps to 1Gbps, I no longer have any problems with apollo and moonlight. Now I can stream 4K 60 fps with no worries! Thank you again :)

**Comment by renaudcerrato** [comment]
I'm not gonna downgrade my host ethernet adapter for Moonlight. I'm instead using a powershell script to force 1Gbe speed through a custom/dummy launcher in Moonlight until that bug is fixed.

There's near zero chance that's an hardware issue : my setup has no problem streaming games in VR to my headset at 500Mbps+ bitrate using Virtual Desktop - streaming through the **same network switches**.

**Comment by bmlrX3k** [comment]
@renaudcerrato can you share your powershell script ?

**Comment by renaudcerrato** [comment]
> [@renaudcerrato](https://github.com/renaudcerrato) can you share your powershell script ?

Sure, [here it is](https://gist.github.com/renaudcerrato/99e423b84e1c0200793467dedad2a0f7).

I manually created 2 "launchers" for Moonshine from Sunshine (within "Applications") named "1Gbps" and "2.5Gbps" that I'm manually launching before and after a gaming session. 


![Image](https://github.com/user-attachments/assets/2c047c30-0590-426d-b3e9-2b090be77ba5)

Make sure "run as administrator" is checked and use that command:

![Image](https://github.com/user-attachments/assets/b490f22f-4c50-4be4-a14a-918386193b19)

`powershell -executionpolicy bypass -File SetAdapter.ps1 -AdapterName YourAdapterName -RegistryKeyword "*SpeedDuplex" -TargetValue 6`

Obviously, this will kill your connection to Sunshine since the adapter will be restarted and you'll get an error, but you'll return to the main screen and you can then launch your game at 1Gbe.

`AdapterName` is optional if there's only one NIC in your computer. The `TargetValue` may be dependent on your NIC : 

- 6 is full-duplex 1Gbe for my NIC.
- 0 is the default mode (auto)

To get those values, open your device manager, open the advanced settings of your NIC manually override the duplex settings and you can run to get the current value:

`Get-NetAdapterAdvancedProperty -RegistryKeyword "*SpeedDuplex"`

or simply run the script without `TargetValue`:

`powershell -executionpolicy bypass -File SetAdapter.ps1 -RegistryKeyword "*SpeedDuplex"`

Output:

```powershell
AdapterName not specified. Searching for adapters that support '*SpeedDuplex'...
Checking which adapters support '*SpeedDuplex'...
Automatically selected adapter 'Thermal' as it's the only one supporting '*SpeedDuplex'.

Operating in Display Mode for adapter 'Thermal', keyword '*SpeedDuplex'.
--------------------------------------------------
Property Details for Adapter: Thermal
--------------------------------------------------
Registry Keyword : *SpeedDuplex
Display Name     : Vitesse et duplex
Display Value    : Négociation automatique
Registry Value(s): 0
--------------------------------------------------

Script finished.
```

**Comment by bmlrX3k** [comment]
Thank you @renaudcerrato !

**Comment by CommonMugger** [comment]
This was driving me insane thinking something was very wrong with my network setup ( there isn't, it's very good). If this was the problem all along, I'll be very sad. I'm going to try it tonight. Thanks @renaudcerrato , you're a god damn hero.

**Comment by MariusBeul** [comment]
In case it helps someone: For me it was solved by removing my old Netgear GS105 switch.

Host <-> GS305 <-> FRITZ!Box 7530 AX <-> **GS105** <-> N100 client

With the 10y old GS105 switch in the pipeline I got 100% frame loss with 30+ Mbits. Without it, 150Mbit 4k/60 FPS is buttery smooth. Maybe the discussed bursts pose a problem for the 15y old chipset in the GS105.

**Comment by FoggyBytes** [comment]
Reading through this issue thread was the only thing that finally helped me understand why my stream was constantly dropping packets and stuttering. The 2.5Gbps (Host) to 1.0Gbps (Client) mismatch and the resulting switch buffer overflow is definitely the culprit for many of us.

Manually throttling the host adapter down to 1.0 Gbps via Windows Device Manager perfectly mitigates the issue, but doing it manually every day was driving me crazy.

Inspired by the discussions here, I decided to build a dedicated workaround tool. I wrote a very lightweight, open-source WPF tray application called **Network Speed Toggle**. 
It sits silently in the Windows system tray and allows you to instantly toggle the Ethernet link speed between `1.0 Gbps` (for Moonlight streaming) and `2.5 Gbps` (for normal use/downloads) with a simple double-click, completely bypassing the UAC prompts.

If anyone here is tired of digging through network adapter properties before every gaming session, I hope this little tool can be useful to you:
https://github.com/FoggyPunk/NetworkSpeedToggle

Thanks to everyone in this thread for investigating and finding the root cause of this network behavior!

**Comment by FoggyBytes** [comment]
> Reading through this issue thread was the only thing that finally helped me understand why my stream was constantly dropping packets and stuttering. The 2.5Gbps (Host) to 1.0Gbps (Client) mismatch and the resulting switch buffer overflow is definitely the culprit for many of us.
> 
> Manually throttling the host adapter down to 1.0 Gbps via Windows Device Manager perfectly mitigates the issue, but doing it manually every day was driving me crazy.
> 
> Inspired by the discussions here, I decided to build a dedicated workaround tool. I wrote a very lightweight, open-source WPF tray application called **Network Speed Toggle**. It sits silently in the Windows system tray and allows you to instantly toggle the Ethernet link speed between `1.0 Gbps` (for Moonlight streaming) and `2.5 Gbps` (for normal use/downloads) with a simple double-click, completely bypassing the UAC prompts.
> 
> If anyone here is tired of digging through network adapter properties before every gaming session, I hope this little tool can be useful to you: https://github.com/FoggyPunk/NetworkSpeedToggle
> 
> Thanks to everyone in this thread for investigating and finding the root cause of this network behavior!

If you're interested, I recommend checking out the new versions I developed today. You can go straight to v1.2.1.
